In [1]:
import math
import statistics
import scipy.stats as stats

In [2]:
results = []
with open("result.txt", "r") as f:
    results = [tuple(map(int, x.strip().split())) for x in f.readlines()]

print(*results, sep="\n")

(56, 400)
(44, 44)
(256, 636)
(14, 14)
(56, 130)
(9, 9)
(566, 6279)
(32, 141)
(16, 16)
(126, 593)
(68, 113)
(16, 16)
(28, 68)
(174, 742)
(39, 61)
(24, 42)
(80, 583)
(30, 172)
(42, 167)
(24, 24)
(27, 62)
(32, 101)
(36, 36)
(116, 1113)
(205, 978)
(32, 48)
(99, 697)
(21, 21)
(8, 8)
(10, 10)
(1172, 3529)
(31, 51)
(219, 379)
(21, 21)
(239, 786)
(1729, 4744)
(75, 121)
(10, 10)
(7, 7)
(260, 1180)
(22, 30)
(167, 798)
(19, 43)
(73, 477)
(29, 29)
(27, 39)
(157, 745)
(660, 1903)
(68, 423)
(110, 296)
(40, 40)
(646, 2841)
(63, 79)
(27, 98)
(30, 95)
(190, 867)
(114, 195)
(65, 85)
(31, 155)
(146, 646)
(100, 383)
(45, 54)
(24, 24)
(105, 483)
(256, 1518)
(56, 145)
(30, 133)
(283, 480)
(185, 417)
(60, 60)
(56, 56)
(118, 321)
(132, 132)
(105, 187)
(69, 385)
(71, 95)
(42, 72)
(140, 388)
(220, 552)
(30, 59)
(77, 77)
(48, 118)
(48, 48)
(24, 24)
(47, 79)
(21, 21)
(248, 724)
(49, 49)
(97, 125)
(6, 6)
(36, 60)
(55, 55)
(35, 55)
(298, 2076)
(24, 24)
(112, 134)
(72, 72)
(88, 120)
(60, 70)
(144, 239)


Analyzing Manhattan cost

In [3]:
manhattan_costs = [x[0] for x in results]
print("Mean:", statistics.mean(manhattan_costs))
print("Standard deviation:", statistics.stdev(manhattan_costs))
print("Max:", max(manhattan_costs))
print("Min:", min(manhattan_costs))

Mean: 124.49
Standard deviation: 227.0569053845936
Max: 1729
Min: 6


Analyzing Misplaced cost

In [4]:
misplaced_costs = [x[1] for x in results]
print("Mean:", statistics.mean(misplaced_costs))
print("Standard deviation:", statistics.stdev(misplaced_costs))
print("Max:", max(misplaced_costs))
print("Min:", min(misplaced_costs))

Mean: 441.56
Standard deviation: 927.9036143780007
Max: 6279
Min: 6


Analyzing (Misplaced - Manhattan)

H0 = 0 = A* search with Manhattan distance heuristic function and Misplaced tile heuristic function is equal.
HA = A* search with Manhattan distance heuristic function and Misplaced tile heuristic function is not equal.

degree of freedom = 100 -1 = 99

In [7]:
diff_manhattan_misplace = [ (x[1] - x[0]) for x in results]
mean = statistics.mean(diff_manhattan_misplace)
std_deviation = statistics.stdev(diff_manhattan_misplace)
std_error = std_deviation/math.sqrt(100)
print("Mean:", mean)
print("Standard deviation:", std_deviation)
print("Max:", max(diff_manhattan_misplace))
print("Min:", min(diff_manhattan_misplace))
print("Standard Error of Mean:", std_error)
print("t-value", (mean-0)/std_error)

Mean: 317.07
Standard deviation: 747.389466913018
Max: 5713
Min: 0
Standard Error of Mean: 74.7389466913018
t-value 4.242366450648694


Let significance value 	α = 0.05
t_critical ~ 1.98 < 4.242
Though reject the null hypothesis.

So, Manhattan heuristic function perform better in than Misplace heuristic function for solving
Modified N puzzle problem